accepted/benchmark-suite.md

```diff
@@ -0,0 +1,796 @@
+
+This benchmark suite has two primary goals:
+
+1. Automatically detecting unintended performance regressions
+
+2. Evaluating the speedup of proposed performance optimizations and overhead of
+   security mitigations
+
+## Automatically Detecting Unintended Performance Regressions
+
+Without dilligent monitoring, unintended regressions will sneak past us. Manual
+review is not sufficient, as the success of continuous integration tests shows
+us. While our testing and fuzzing infrastructure will generally catch
+correctness regressions, we have nothing in place for catching performance
+regressions.
+
+The benchmark suite should be run once per day (or as often as is feasible,
+ideally it would run for every commit) and report any regressions in execution
+time or memory overhead.
+
+## Evaluating Proposed Performance Optimizations and Security Mitigations
+
+The decision of whether to include a proposed optimization must balance
+maintainability, compile times, and runtime improvement. Maintainability is
+subjective, but our new benchmark should provide quantitative data for the
+latter two: how does the new optimization affect compile times, and what kind of
+speed ups can we expect? Does it shrink emitted code size or impose memory
+overhead? The answers to these questions will help us evaluate a proposed
+performance optimization.
+
+Similarly, these same types of questions appear when evaluating a security
+mitigation. How much overhead does, for example, emitting
+[retpolines](https://software.intel.com/security-software-guidance/api-app/sites/default/files/Retpoline-A-Branch-Target-Injection-Mitigation.pdf)
+add to WebAssembly execution time?
+
+Experiments like these are invaluable for engineers and researchers working with
+Wasmtime and Cranelift.
+
+## Nongoal: Creating a General-Purpose WebAssembly Benchmark Suite
+
+It is also worth mentioning this explicit nongoal: we do not intend to develop a
+general-purpose WebAssembly benchmark suite, used to compare between different
+WebAssembly compilers and runtimes. We don't intend to trigger a WebAssembly
+benchmarking war, reminiscent of JavaScript benchmarking wars in Web
+browsers. Doing so would make the benchmark suite's design high stakes, because
+engineers would be incentivized to game the benchmarks, and would additionally
+impose cross-engine portability constraints on the benchmark runner. We only
+intend to compare the performance of various versions of Wasmtime and Cranelift,
+where we don't need the cross-engine portability in the benchmark runner, and
+where gaming the benchmarks isn't incentivized.
+
+Furthermore, general-purpose WebAssembly benchmarking must include WebAssembly
+on the Web. Doing that well requires including interactions with the rest of the
+Web browser: JavaScript, rendering, and the DOM. Building and integrating a full
+Web browser is overkill for our purposes, and represents significant additional
+complexity that we would prefer to avoid.
```

Comment:
Reading through this RFC again, one thing that occurs to me is that it doesn't clarify whether "microbenchmarks" are in scope or not. On one hand, we might say that micro-benchmarks are in indeed in scope, specifically because we're not trying to build a General-Purpose benchmark suite, but instead just something to collect data to help identify performance changes over time. On the other hand, some parts of this proposal talk about a desire for a representative corpus. Could you clarify the intended stance on microbenchmarks?

Reply:
I think my position is summarized in my response to Johnnie up thread:

> Micro-benchmarks have their place, and we can include some as we find blind spots and want to target specific things (like Lars also mentioned) but I think it makes sense, especially as we create the initial set of candidates, to focus on real programs.

Multiple people have mentioned the desire to run microbenchmarks, and if we get value out of having them, I don't want to stand in the way for ideological reasons. As long as the microbenchmarks use the same benchmarking API/protocol that the rest of the corpus uses, supporting them should be straightforward and shouldn't require any additional engineering effort.